base_layer/core/src/transaction.rs

```diff
@@ -84,10 +186,210 @@ pub struct TransactionKernel {
+}
+
+/// Implementation of the transaction kernel
+impl<'a> TransactionKernel<'a> {
+    /// Creates an empty transaction kernel
+    pub fn empty() -> TransactionKernel<'a> {
+        TransactionKernel { features: KernelFeatures::empty(), fee: 0, lock_height: 0, excess: None, excess_sig: None }
+    }
+
+    /// Build a transaction kernel with the provided fee
+    pub fn with_fee(self, fee: u64) -> TransactionKernel<'a> {
+        TransactionKernel { fee, ..self }
+    }
+
+    /// Build a transaction kernel with the provided lock height
+    pub fn with_lock_height(self, lock_height: u64) -> TransactionKernel<'a> {
+        TransactionKernel { lock_height, ..self }
+    }
+}
+
+/// Implement the canonical hashing function for TransactionKernel for use in ordering
+impl<'a> Hashable for TransactionKernel<'a> {
+    type Hasher = Blake256;
+
+    fn hash(&self) -> Vec<u8> {
+        let mut hasher = Self::Hasher::new();
+        hasher.input(vec![self.features.bits]);
+        let mut fee_lock_height = Vec::new();
+        fee_lock_height.write_u64::<LittleEndian>(self.fee).unwrap();
+        fee_lock_height.write_u64::<LittleEndian>(self.lock_height).unwrap();
+        hasher.input(fee_lock_height);
+        if self.excess.is_some() {
+            hasher.input(self.excess.unwrap().to_bytes());
+        }
+        if self.excess_sig.is_some() {
+            hasher.input(self.excess_sig.unwrap().get_signature().to_bytes());
+        }
+        hasher.result().to_vec()
+    }
+}
+
+hashable_ord!(TransactionKernel);
+
+/// A transaction which consists of a kernel offset and an aggregate body made up of inputs, outputs and kernels.
+pub struct Transaction<'a> {
+    /// This kernel offset will be accumulated when transactions are aggregated to prevent the "subset" problem where
+    /// kernels can be linked to inputs and outputs by testing a series of subsets and see which produce valid
+    /// transactions
+    pub offset: Option<BlindingFactor>,
+    /// The constituents of a transaction which has the same structure as the body of a block.
+    pub body: AggregateBody<'a>,
+    /// reference to the Base point used in the commitments in this transaction
+    pub base: &'a Base,
+}
+
+impl<'a> Transaction<'a> {
+    /// Create an empty transaction
+    pub fn empty(base: &'a Base) -> Transaction<'a> {
+        Transaction { base, offset: None, body: AggregateBody::empty() }
+    }
+
+    /// Create a new transaction from the provided inputs, outputs, kernels and offset
+    pub fn new(
+        base: &'a Base,
+        inputs: Vec<TransactionInput<'a>>,
+        outputs: Vec<TransactionOutput<'a>>,
+        kernels: Vec<TransactionKernel<'a>>,
+        offset: Option<BlindingFactor>,
+    ) -> Transaction<'a>
+    {
+        Transaction { base, offset, body: AggregateBody::new(inputs, outputs, kernels) }
+    }
+
+    /// Update the offset of an existing transaction
+    pub fn with_offset(self, offset: BlindingFactor) -> Transaction<'a> {
```

Comment:
Also seems like this should be in a `TransactionBuilder` struct

Reply:
agreed.